src/compiler/checker.ts

```diff
@@ -5351,17 +5351,39 @@ module ts {
                 }
                 // Target type is type of prototype property
                 let prototypeProperty = getPropertyOfType(rightType, "prototype");
-                if (!prototypeProperty) {
-                    return type;
+                if (prototypeProperty) {
+                    let targetType = getTypeOfSymbol(prototypeProperty);
+                    if (targetType !== anyType) {
+                        // Narrow to target type if it is a subtype of current type
+                        if (isTypeSubtypeOf(targetType, type)) {
+                            return targetType;
+                        }
+                        // If current type is a union type, remove all constituents that aren't subtypes of target type

```

Comment:
Nit: "If the current type is a union type, remove all constituents that aren't subtypes of the target."